### ISSUE TITLE
[clang-format] qualifier alignment setting is not respected on references to template type parameter

### ISSUE THREAD

**Issue by alimpfard** [llvm/llvm-project#54730]
Given the following .clang-format (and clang-format 14.0.0)
```yaml
---
Language: Cpp
QualifierAlignment: Right
```

the following snippet is not formatted correctly (it remains unchanged)
```c++
template <typename T> using Foo = const volatile T;
```

however the following does get replaced with the qualifiers moving to the right:
```c++
template <typename T> using Foo = const volatile int;
// turns into:
// template <typename T> using Foo = int const volatile;
```

**Comment by alimpfard** [comment]
The symptoms don't immediately seem to be the same (reference to typename vs argument), but I haven't tested with 28bb040ded83b7cffdd6e6102080ccbe7935dbfe applied.

**Comment by mydeveloperday** [comment]
Hmm... you may be correct the follow producer fails

```
TEST_F(QualifierFixerTest, TemplatesRightVolatile) {
  FormatStyle Style = getLLVMStyle();
  Style.QualifierAlignment = FormatStyle::QAS_Right;
  Style.QualifierOrder = {"type", "volatile", "const"};

  verifyFormat("template <typename T> using Foo = int volatile const;",
               "template <typename T> using Foo = const volatile int;", Style);
  verifyFormat("template <typename T> using Foo = T volatile const;",
               "template <typename T> using Foo = const volatile T;", Style);
}
```

**Comment by mydeveloperday** [comment]
This is really a special case that is not covered, take the following tests, the last one will fail because it does't have a "variable

```c
  verifyFormat("int volatile const a;", "const volatile int a;", Style);
  verifyFormat("int volatile const Foo;", "const volatile int Foo;", Style);
  verifyFormat("int volatile const T;", "const volatile int T;", Style);
  verifyFormat("T volatile const;", "const volatile T;", Style);
```

but this would work,

```c
verifyFormat("T volatile const a;", "const volatile T a;", Style);
```

To the same extent, this won't work because it doesn't see the T as a type but as an identifier

```c
 verifyFormat("T volatile const a;", "const volatile T a;", Style);
```

we air on the side of caution for now, but in time we might find a way of doing this. Qualifier alignment is going to work much better with simple types. For us to have more confidence to make changes we may need to introduce the ability to give clang-format a little more type information to certain identifiers are types something like

```yaml
QUALIFER_TYPES= [ 'T', 'Foo' ]
```

**Comment by alimpfard** [comment]
:thinking: Shouldn't clang-format be able to infer the fact that `T` is a type based on the fact that the "expression" it's in is a type specifier?
it appears to me that clang-format does have _some_ concept of type-"expression" vs value-"expression" as these are formatted differently:
```c++
typedef T * V;    // typedef T* V;
typedef 4 * V;    // typedef 4 * V;
decltype(T * V);  // decltype(T * V); 
```

**Comment by rymiel** [comment]
This should be fixed by cd7ab4b5c1684dcc60de027700177adfa096b98c
